hack/generator/pkg/codegen/pipeline_create_storage_types.go

```diff
@@ -0,0 +1,212 @@
+	result.VisitArmType = factory.visitArmType
+	//result.VisitResourceType = factory.visitResourceType
+
+	return result
+}
+
+type StorageTypeFactory struct {
+	types astmodel.Types
+}
+
+func (factory *StorageTypeFactory) visitTypeName(_ *astmodel.TypeVisitor, name astmodel.TypeName, ctx interface{}) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+
+	// Resolve the type name to the actual referenced type
+	actualDefinition, actualDefinitionFound := factory.types[name]
+
+	// Check for property specific handling
+	if vc.property != nil && actualDefinitionFound {
+
+		if et, ok := actualDefinition.Type().(*astmodel.EnumType); ok {
+			// Property type refers to an enum, so we use the base type instead
+			return et.BaseType(), nil
+		}
+	}
+
+	// Map the type name into our storage namespace
+	visitedName, err := factory.mapTypeName(name)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to map name into storage namespace")
+	}
+
+	return visitedName, nil
+}
+
+func (factory *StorageTypeFactory) visitObjectType(
+	visitor *astmodel.TypeVisitor,
+	object *astmodel.ObjectType,
+	ctx interface{},
+) (astmodel.Type, error) {
+	vc := ctx.(StorageTypesVisitorContext)
+	oc := vc.forObject(object)
+
+	var errs []error
+	properties := object.Properties()
+	for i, prop := range properties {
+		propertyType, err := visitor.Visit(prop.PropertyType(), oc.forProperty(prop))
+		if err != nil {
+			errs = append(errs, err)
+		} else {
+			p := factory.makeStorageProperty(prop, propertyType)
+			properties[i] = p
+		}
+	}
+
+	if len(errs) > 0 {
+		err := kerrors.NewAggregate(errs)
+		return nil, err
+	}
+
+	ot := astmodel.NewObjectType().WithProperties(properties...)
+	return astmodel.MakeStorageType(ot), nil
+}
+
+func (factory *StorageTypeFactory) makeStorageProperty(prop *astmodel.PropertyDefinition, propertyType astmodel.Type) *astmodel.PropertyDefinition {
+	p := prop.WithType(propertyType).
+		MakeOptional().
+		WithoutValidation().
+		WithDescription("")
+
+	// If the property is *string we can simplify it to string
+	// This makes other generated code simpler
+	if ot, ok := p.PropertyType().(*astmodel.OptionalType); ok {
+		if ot.Element().Equals(astmodel.StringType) {
+			p = p.WithType(astmodel.StringType)
+		}
+	}
+
+	return p
+}
+
+// mapTypeName maps an existing type name into the right package for the matching storage type
+// Returns the original instance for any type that does not need to be mapped to storage
+func (factory *StorageTypeFactory) mapTypeName(name astmodel.TypeName) (*astmodel.TypeName, error) {
+	if !name.PackageReference.IsLocalPackage() {
+		// Don't need to map non-local packages
+		return &name, nil
+	}
+
+	storagePackage, err := astmodel.CreateStoragePackageReference(name.PackageReference)
+	if err != nil {
+		return nil, err
+	}
+
+	newName := astmodel.MakeTypeName(storagePackage, name.Name())
+	return &newName, nil
+}
+
+func (factory *StorageTypeFactory) visitArmType(
```

Comment:
Does this function ever get called?

You already skipped arm types in the loop, so how would a visit function run into an ARM type? I suppose it would only be possible if a non-ARM type had a property that was an ARM type (AFAIK that doesn't happen?)

is that actually an error condition?

Reply:
It happens if an ArmType gets wrapped by something else. Won't happen via a property as we do different things there - but mostly this is defensive.

Reply:
Rather than be defensive here wouldn't it just be better to error?

It seems to me that "storage types shall not have/consume ARM types" is a fundamental truth, yes?

Reply:
The visitor walks the entire tree; if we don't truncate the walk here, we'd end up creating a storage type based on the object type wrapped by the ArmType - which we don't want/need.